src/NLog/GlobalDiagnosticsContext.cs

```diff
@@ -65,13 +62,29 @@ public static void Set(string item, string value)
         /// <param name="item">Item name.</param>
         /// <param name="value">Item value.</param>
         public static void Set(string item, object value)
+        {
+            SetInternal(item, value);
+        }
+
+        private static void SetInternal(string item, object value)
         {
             lock (_dict)
             {
-                GetWritableDict(_dictReadOnly != null && !_dict.ContainsKey(item))[item] = value;
+                if (ContainsItemInternal(item))
```

Comment:
Missing a not-operator !. When the key already exist in the dictionary then it is safe to override its value, because it doesn't have to resize container.

Reply:
oops, fixed